### PR TITLE
[Frontend] support matryoshka representation / support embedding API dimensions

### LINKED ISSUE #15465: [Feature]: Embedding API dimensions  is currently not supported.
### 🚀 The feature, motivation and pitch

I noticed that after deploying the embedding model using vllm, the requests do not fully match the OpenAI format.  
This issue has always been there:
```
{
  "object": "error",
  "message": "dimensions is currently not supported",
  "type": "BadRequestError",
  "param": null,
  "code": 400
}
```

### Alternatives

My script:
```
vllm serve "/workspace/share_data/base_llms/bce-embedding-base_v1" \
    --served-model-name "bce-embedding-base_v1" \
    --task "embedding" \
    --trust-remote-code \
    --host "0.0.0.0" \
    --port 8000 \
    --dtype auto \
    --gpu-memory-utilization 0.4 \
    --kv-cache-dtype auto \
    --enable-prefix-caching \
    --tensor-parallel-size 1 \
    --max-num-seqs 256
```
vllm version: 0.7.2

### Additional context

I hope this feature can be added as soon as possible, as it will be immensely helpful for building the knowledge base.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by noooop**
## Summary

[Matryoshka Embeddings](https://sbert.net/examples/sentence_transformer/training/matryoshka/README.html#matryoshka-embeddings) or [Matryoshka Representation Learning (MRL)](https://arxiv.org/abs/2205.13147) is a technique used in training embedding models. It allows user to trade off between performance and cost.

**Not all embeddings models support MRL. Changing the output dimension for models that do not support MRL will lead to poor results. vllm returns an error for requests that attempt to change the output dimension of an unsupported MRL model.**

```
  raise ValueError(
      f'Model "{model_config.served_model_name}" does not '
      f'support matryoshka representation, '
      f'changing output dimensions will lead to poor results.')
```

We hope that the open source community will adopt the terms “is_matryoshka ” or “matryoshka_dimensions ” to denote whether a model is compatible with Matryoshka Representation Learning (MRL).

### offline
```
python -m examples.offline_inference.embed_matryoshka_fy
```
## online 

```
vllm serve jinaai/jina-embeddings-v3 --trust-remote-code
```

```
curl http://127.0.0.1:8000/v1/embeddings \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
    "input": "The capital of Brazil is Brasilia.",
    "model": "jinaai/jina-embeddings-v3",
    "encoding_format": "float",
    "dimensions": 1
  }'
```

expected output
```
{"id":"embd-0aab28c384d348c3b8f0eb783109dc5f","object":"list","created":1744195454,"model":"jinaai/jina-embeddings-v3","data":[{"index":0,"object":"embedding","embedding":[-1.0]}],"usage":{"prompt_tokens":10,"total_tokens":10,"completion_tokens":0,"prompt_tokens_details":null}}
```

FIX #15465

### CODE REVIEW

**Comment by noooop** [comment]
@DarkLight1337


Currently the following tests can be passed locally:

- pytest tests/models/embedding/language/test_jina.py`
- pytest tests/entrypoints/openai/test_embedding_dimensions.py
- python examples/offline_inference/embed_matryoshka_fy.py


potential problems:
- Variable naming and docstrings
- offline PoolingParams verify
- test_embedding_dimensions.py merge into tests/entrypoints/openai/test_embedding.py
- Is online (openai server) accuracy testing required?

File: tests/entrypoints/openai/test_embedding_dimensions.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+from collections import namedtuple
+
+import openai
+import pytest
+
+from vllm.entrypoints.openai.protocol import EmbeddingResponse
+
+from ...utils import RemoteOpenAIServer
+
+ModelInfo = namedtuple('ModelInfo', ['name', 'is_matryoshka'])
```
Comment (DarkLight1337) [inline tests/entrypoints/openai/test_embedding_dimensions.py:15]: Use `typing.NamedTuple`

File: vllm/model_executor/layers/pooler.py [hunk 2]
```diff
@@ -217,14 +218,38 @@ def __init__(self, *, normalize: bool, softmax: bool) -> None:
...
+                   for pooling_param in pooling_params):
+            # for matryoshka representation
+            pooled_data_list = []
+            for vecs, pooling_param in zip(pooled_data, pooling_params):
+                assert isinstance(vecs, torch.Tensor)
+
+                if pooling_param.dimensions is not None:
+                    # matryoshka representation, always normalize
+                    vecs = vecs[:pooling_param.dimensions]
+                    vecs = F.normalize(vecs, p=2, dim=0)
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:238]: We should disallow `self.normalize=False` in this case

File: vllm/pooling_params.py [hunk 3]
```diff
@@ -12,14 +15,31 @@ class PoolingParams(
...
+    dimensions: Optional[int] = None
     additional_data: Optional[Any] = None
 
     def clone(self) -> "PoolingParams":
         """Returns a deep copy of the PoolingParams instance."""
-        return PoolingParams(additional_data=self.additional_data)
+        return PoolingParams(dimensions=self.dimensions,
+                             additional_data=self.additional_data)
+
+    def verify(self, model_config: "ModelConfig") -> None:
```
Comment (noooop) [inline vllm/pooling_params.py:31]: Finally I chose to add a verify method to PoolingParams.


Is there any way to get the dimensions of the output embeddings so I can verify the upper limit of the dimensions?

Comment (noooop) [inline vllm/pooling_params.py:31]: Finally I chose to add a verify method to PoolingParams.

Is there any way to get the dimensions of the output embeddings so I can verify the upper limit of the dimensions?

File: vllm/pooling_params.py [hunk 4]
```diff
@@ -12,14 +15,31 @@ class PoolingParams(
...
+        if not model_config.is_matryoshka:
+            if self.dimensions is not None:
+                raise ValueError(
+                    f'Model "{model_config.served_model_name}" does not '
+                    f'support matryoshka representation, '
+                    f'changing output dimensions will lead to poor results.')
+        else:
+            if isinstance(self.dimensions, int) and self.dimensions < 1:
+                raise ValueError("Dimensions must be greater than 0")
```
Comment (noooop) [inline vllm/pooling_params.py:40]: Finally I chose to add a verify method to PoolingParams.

Is there any way to get the dimensions of the output embeddings so I can verify the upper limit of the dimensions?

File: vllm/model_executor/layers/pooler.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py]: This isn't resolved yet

**Comment by noooop** [comment]
@DarkLight1337 

Pass is_matryoshka to PoolerHead via pooler_config to increase logic clarity

Now the logic is:

Fully controlled by is_matryoshka, always normalize when is_matryoshka

File: vllm/model_executor/layers/pooler.py [hunk 5]
```diff
@@ -211,20 +228,47 @@ def extract_states(
...
-
-    def forward(self, pooled_data: Union[list[torch.Tensor], torch.Tensor]):
-        if self.normalize:
-            if isinstance(pooled_data, list):
-                pooled_data = [
-                    F.normalize(data, p=2, dim=1) for data in pooled_data
-                ]
-            else:
-                pooled_data = F.normalize(pooled_data, p=2, dim=1)
+        self.is_matryoshka = is_matryoshka
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:237]: I think we should disallow the combination `normalize=False, is_matryoshka=True`. We should adjust the default value so that `normalize=True` if `is_matryoshka` is set by the user

File: vllm/model_executor/layers/pooler.py [hunk 6]
```diff
@@ -211,20 +228,47 @@ def extract_states(
...
+            # for matryoshka representation, always do normalize
+            assert len(pooled_data) == len(pooling_params)
+            assert self.is_matryoshka is True
+
+            pooled_data_list = []
+            for i, pooling_param in enumerate(pooling_params):
+                vecs = pooled_data[i]
+
+                if pooling_param.dimensions is not None:
+                    vecs = vecs[:pooling_param.dimensions]
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:267]: I think we should move this before normalize. That way we don't have to write normalization code twice

**Comment by noooop** [comment]
@DarkLight1337 

Split the normalize and the change the output dimension

how about now

File: vllm/model_executor/layers/pooler.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:236]: How about this?

**Comment by noooop** [comment]
@DarkLight1337 

I have closed all previous versions of conversation. 

Do you have any suggestions to the latest version?

File: vllm/model_executor/layers/pooler.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:236]: I mean that we should not let user set `normalize=False`

**Comment by noooop** [comment]
@DarkLight1337 

how about “overwrite normalize in _init_pooler_config”

**Comment by DarkLight1337** [comment]
> @DarkLight1337
> 
> how about “overwrite normalize in _init_pooler_config”

I think we should make this a user-facing error instead of silently overwriting the user's configuration.

File: vllm/config.py [hunk 7]
```diff
@@ -500,16 +500,20 @@ def _init_pooler_config(
         if self.runner_type == "pooling":
             user_config = override_pooler_config or PoolerConfig()
 
+            if self.is_matryoshka and user_config.normalize is None:
+                user_config.normalize = True
+
```
Comment (DarkLight1337) [inline vllm/config.py:505]: Check this after pooling conifg

**Comment by noooop** [comment]
@DarkLight1337


raise ValueError when is matryoshka and normalize being disabled

File: vllm/pooling_params.py (hunk 4 above)
Comment (DarkLight1337) [inline vllm/pooling_params.py:40]:
```suggestion
        if self.dimensions is not None:
            if not model_config.is_matryoshka:
                raise ValueError(
                    f'Model "{model_config.served_model_name}" does not '
                    f'support matryoshka representation, '
                    f'changing output dimensions will lead to poor results.')

            if self.dimensions < 1:
                raise ValueError("Dimensions must be greater than 0")
```

Simplify the check a bit

File: vllm/model_executor/layers/pooler.py [hunk 8]
```diff
@@ -217,14 +218,35 @@ def __init__(self, *, normalize: bool, softmax: bool) -> None:
...
+
+                pooled_data_list.append(vecs)
+            pooled_data = pooled_data_list
+
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:242]:
```suggestion
        dimensions_list = [
            pooling_param.dimensions
            for _, pooling_param in pooling_metadata.seq_groups
        ]

        if any(d is not None for d in dimensions_list):
            assert len(pooled_data) == len(dimensions_list)

            pooled_data = [
                vecs if d is None else vecs[:d]
                for vecs, d in zip(pooled_data, dimensions_list)
            ]

```

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks for bearing with me!

**Comment by noooop** [comment]
Thanks for reviewing


I think I should submit more code to the open source community to improve my coding skills.

**Comment by DarkLight1337** [comment]
I think we can also update the docs for Pooling Models to tell users how to use `is_matryoshka`, but that can be done in another PR if you want

**Comment by noooop** [comment]
> We should also update the docs for Pooling Models to tell users how to use `is_matryoshka`, but that can be done in another PR if you want.

I will try.

**Comment by noooop** [comment]
@DarkLight1337 


CI is stuck


Every time pr go to the CI stage, I don’t know what to do

**Comment by DarkLight1337** [comment]
Force merging

**Comment by noooop** [comment]
> Force merging

QvQ
